### ISSUE TITLE
bcache conversion /dev/mapper/synthetic read errors 

### ISSUE THREAD

**Issue by RelativePrime** [g2p/blocks#6]
I'm having trouble getting blocks to-bcache to create a device I am able to subsequently boot to. 

The target system in OpenSuse 13.1 with bcache-tools installed. I'm then working from an Ubuntu Saucy LiveCD to perform the conversion (this just due to a bug with python3 in the current OpenSuse live CD). The caching device appears to be successfully created with:

```
root@ubuntu:~# make-bcache -C /dev/sdb --wipe
UUID:            f5932af1-dba9-472c-876e-346f68edc5e2
Set UUID:        1262d28f-6b93-4f09-90fe-ca85e554cf86
version:        0
nbuckets:        228946
block_size:        1
bucket_size:        1024
nr_in_set:        1
nr_this_dev:        0
first_bucket:        1
```

Proceeding with the conversion then results in:

```
root@ubuntu:~# blocks to-bcache /dev/mapper/system-root --join 1262d28f-6b93-4f09-90fe-ca85e554cf86
Will shrink the filesystem (btrfs) by 4194304 bytes
Command ['dmsetup', 'remove', '--noudevsync', '--', 'synthetic-5c7be7ba-5b9a-11e3-ad03-842b2b9e724e'] has failed with status 1
Standard output:

Standard error:
device-mapper: remove ioctl on synthetic-5c7be7ba-5b9a-11e3-ad03-842b2b9e724e failed: Device or resource busy
Command failed

Command ['dmsetup', 'remove', '--noudevsync', '--', 'rozeros-5c7b9b8e-5b9a-11e3-ad03-842b2b9e724e'] has failed with status 1
Standard output:

Standard error:
device-mapper: remove ioctl on rozeros-5c7b9b8e-5b9a-11e3-ad03-842b2b9e724e failed: Device or resource busy
Command failed

Copying the bcache superblock... ok
  /dev/mapper/synthetic-5c7be7ba-5b9a-11e3-ad03-842b2b9e724e: read failed after 0 of 4096 at 42949607424: Input/output error
  /dev/mapper/synthetic-5c7be7ba-5b9a-11e3-ad03-842b2b9e724e: read failed after 0 of 4096 at 42949664768: Input/output error
  mpath major 251 is not dm major 252.
Loading LVM metadata... ok
Rotating the last extent to be the first... ok
```

I'm unclear on the  cause of the read errors on /dev/synthetic. The same occurs with:

```
root@ubuntu:~# vgscan 
  Reading all physical volumes.  This may take a while...
  /dev/mapper/synthetic-5c7be7ba-5b9a-11e3-ad03-842b2b9e724e: read failed after 0 of 4096 at 42949607424: Input/output error
  /dev/mapper/synthetic-5c7be7ba-5b9a-11e3-ad03-842b2b9e724e: read failed after 0 of 4096 at 42949664768: Input/output error
  Found volume group "system" using metadata type lvm2
```

After rebooting the root device is not found, regardless of if the UUID or root=/dev/bcache0 is used for the grub2 boot arg.

Some additional information which may help:

```
root@ubuntu:~# lsblk -o NAME,MAJ:MIN,RM,SIZE,TYPE,FSTYPE,MOUNTPOINT,UUID
NAME                                                    MAJ:MIN RM   SIZE TYPE FSTYPE    MOUNTPOINT UUID
sda                                                       8:0    0 465.8G disk                      
├─sda1                                                    8:1    0   399M part ext4                 0da71f87-ba24-462d-93ab-3420a49ca7c5
└─sda2                                                    8:2    0 465.4G part LVM2_memb            8G026Q-gJeY-rTbI-xX1z-0uZD-8ZZD-g8BhKf
  ├─system-root (dm-0)                                  252:0    0    40G lvm                       
  │ └─bcache1                                           251:1    0    40G disk btrfs                e6285c09-b1aa-48a8-b92f-e54825ab7024
  └─system-swap (dm-1)                                  252:1    0     4G lvm  swap                 fde68d05-a80b-449d-94e8-69250bc14034
sdb                                                       8:16   0 111.8G disk                      
├─bcache0                                               251:0    0    40G disk                      
└─bcache1                                               251:1    0    40G disk btrfs                e6285c09-b1aa-48a8-b92f-e54825ab7024
sr0                                                      11:0    1   883M rom  iso9660   /cdrom     
rozeros-5c7b9b8e-5b9a-11e3-ad03-842b2b9e724e (dm-2)     252:2    0    40G dm                        
└─synthetic-5c7be7ba-5b9a-11e3-ad03-842b2b9e724e (dm-3) 252:3    0    40G dm                        
  └─bcache0                                             251:0    0    40G disk                      
loop0                                                     7:0    0 842.7M loop squashfs  /rofs      
loop1                                                     7:1    0     4M loop                      
└─synthetic-5c7be7ba-5b9a-11e3-ad03-842b2b9e724e (dm-3) 252:3    0    40G dm                        
  └─bcache0                                             251:0    0    40G disk


root@ubuntu:~# cat /etc/lvm/backup/system 
# Generated by LVM2 version 2.02.98(2) (2012-10-15): Mon Dec  2 21:40:28 2013

contents = "Text Format Volume Group"
version = 1

description = "Created *after* executing 'vgscan'"

creation_host = "ubuntu"    # Linux ubuntu 3.11.0-12-generic #19-Ubuntu SMP Wed Oct 9 16:20:46 UTC 2013 x86_64
creation_time = 1386020428    # Mon Dec  2 21:40:28 2013

system {
    id = "wP6Ta4-bdJb-YjAX-HkZC-g0sH-wWnC-beDh8U"
    seqno = 6
    format = "lvm2" # informational
    status = ["RESIZEABLE", "READ", "WRITE"]
    flags = []
    extent_size = 8192        # 4 Megabytes
    max_lv = 0
    max_pv = 0
    metadata_copies = 0

    physical_volumes {

        pv0 {
            id = "8G026Q-gJeY-rTbI-xX1z-0uZD-8ZZD-g8BhKf"
            device = "/dev/sda2"    # Hint only

            status = ["ALLOCATABLE"]
            flags = []
            dev_size = 975953920    # 465.371 Gigabytes
            pe_start = 2048
            pe_count = 119134    # 465.367 Gigabytes
        }
    }

    logical_volumes {

        root {
            id = "AZMmic-xfZP-g5iv-iBaJ-Nuin-x3rp-6JOUyP"
            status = ["READ", "WRITE", "VISIBLE"]
            flags = []
            creation_host = "linux"
            creation_time = 1385037444    # 2013-11-21 12:37:24 +0000
            segment_count = 2

            segment1 {
                start_extent = 0
                extent_count = 3    # 12 Megabytes

                type = "striped"
                stripe_count = 1    # linear

                stripes = [
                    "pv0", 10237
                ]
            }
            segment2 {
                start_extent = 3
                extent_count = 10237    # 39.9883 Gigabytes

                type = "striped"
                stripe_count = 1    # linear

                stripes = [
                    "pv0", 0
                ]
            }
        }

        swap {
            id = "8JogqY-fkfl-N19f-xtlT-6RrC-dLbE-5Ed1i8"
            status = ["READ", "WRITE", "VISIBLE"]
            flags = []
            creation_host = "linux"
            creation_time = 1385037444    # 2013-11-21 12:37:24 +0000
            segment_count = 1

            segment1 {
                start_extent = 0
                extent_count = 1024    # 4 Gigabytes

                type = "striped"
                stripe_count = 1    # linear

                stripes = [
                    "pv0", 10240
                ]
            }
        }
    }
}
```

I could just dd wipe partition tables on both devices and try yet again, but it seems something else is involved here, or there is something I'm missing.

I would appreciate any suggestions you might have - I'd love to get bcache working on my new Suse 13.1 install.  

Cheers,

**Comment by qpfiffer** [comment]
I'm having a similar issue.

```
quinlan@tired ~ $ sudo /home/quinlan/.virtualenvs/blocks/bin/blocks to-bcache --join=c4f68165-bf87-4d29-8e8e-3f3338deb3f9 /dev/tired/media 
Enter passphrase for /dev/tired/media: 
The filesystem (ext4) leaves enough room, no need to shrink it
device-mapper: remove ioctl on cleartext-4f52858a-6153-11e3-b26a-bc5ff4ceff9f failed: Device or resource busy
device-mapper: reload ioctl on synthetic-536b5e26-6153-11e3-b26a-bc5ff4ceff9f failed: Invalid argument
Command failed
Traceback (most recent call last):
  File "/home/quinlan/.virtualenvs/blocks/bin/blocks", line 9, in <module>
    load_entry_point('blocks==0.1.4', 'console_scripts', 'blocks')()
  File "/home/quinlan/.virtualenvs/blocks/src/blocks/blocks/__main__.py", line 2002, in script_main
    sys.exit(main())
  File "/home/quinlan/.virtualenvs/blocks/src/blocks/blocks/__main__.py", line 1668, in main
    return args.action(args)
  File "/home/quinlan/.virtualenvs/blocks/src/blocks/blocks/__main__.py", line 1751, in cmd_to_bcache
    device=device, debug=args.debug, progress=progress, join=args.join)
  File "/home/quinlan/.virtualenvs/blocks/src/blocks/blocks/__main__.py", line 1500, in lv_to_bcache
    synth_bdev = make_bcache_sb(pe_size, data_size, join)
  File "/home/quinlan/.virtualenvs/blocks/src/blocks/blocks/__main__.py", line 1472, in make_bcache_sb
    with synth_device(bsb_size, data_size) as synth_bdev:
  File "/usr/lib64/python3.3/contextlib.py", line 48, in __enter__
    return next(self.gen)
  File "/home/quinlan/.virtualenvs/blocks/src/blocks/blocks/__main__.py", line 1264, in synth_device
    exit_stack=st)
  File "/home/quinlan/.virtualenvs/blocks/src/blocks/blocks/__main__.py", line 141, in mk_dm
    assert proc.returncode == 0
AssertionError
```

I'd really like to get this working as well.

**Comment by g2p** [comment]
@qpfiffer, someone resolved your situation, look at #8.

@RelativePrime, sorry for not getting back to you earlier.
There's a lot that could be said about these error messages, but ultimately they are innocuous.

> Something is holding open the synthetic, virtual device that's used to build the superblock, which prevents blocks from tearing it down. That provokes extra read errors when lvm peeks at the synthetic device later (which is unavoidable with current dm devices; blocks needs a device that will return zeros on reads and fail on writes, unfortunately the [best that can be done with current kernels](https://www.redhat.com/archives/dm-devel/2013-March/msg00011.html) is a device that errors on reads and writes, and blocks is still prevented from writing out of bounds).

tl;dr, blocks really did complete successfully.
However, bcache-tools doesn't have support for OpenSuse's initramfs. The tarball provides hooks for Arch, Dracut, and Debian/Ubuntu, so it shouldn't be too hard to port; this example is the simplest:
https://github.com/g2p/bcache-tools/blob/master/initcpio/install

You can verify that the initramfs contains everything required without rebooting, using lsinitramfs or something like `zcat /boot/initrd.img-<version> |cpio -t`.

**Comment by g2p** [comment]
Closing; the main issue is https://github.com/g2p/bcache-tools/issues/6.
